dcplib/etl/__init__.py

```diff
@@ -0,0 +1,128 @@
+                content_type_patterns=None, filename_patterns=None, max_workers=512):
+        if query is None:
+            query = self.default_bundle_query
+        if content_type_patterns is None:
+            content_type_patterns = self.default_content_type_patterns
+        os.makedirs("files", exist_ok=True)
+        os.makedirs("bundles", exist_ok=True)
+        f2f_futures, ff_futures = [], []
+        bundles_complete, percent_complete = 0, -1
+        with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
+            dss = hca.dss.DSSClient(swagger_url=f"{self.dss_host}/swagger.json")
+            total_bundles = dss.post_search(es_query=query, replica="aws")["total_hits"]
+            logging.info("Scanning %s bundles", total_bundles)
+            for b in dss.post_search.iterate(es_query=query, replica="aws", per_page=500):
+                bundle_uuid, bundle_version = b["bundle_fqid"].split(".", 1)
+                f2f_futures.append(executor.submit(self.get_files_to_fetch_for_bundle,
+                                                   bundle_uuid, bundle_version, content_type_patterns))
+
+                if len(f2f_futures) > max_workers * 2:
+                    bundles_complete += self.enqueue_bundle_manifests(percent_complete, bundles_complete, total_bundles,
+                                                                      f2f_futures, ff_futures, executor)
+                    f2f_futures.clear()
+
+                if int((bundles_complete / total_bundles) * 100) > percent_complete:
+                    percent_complete = int((bundles_complete / total_bundles) * 100)
+                    logger.info("%d%% complete (%d/%d bundles)", percent_complete, bundles_complete, total_bundles)
+            bundles_complete += self.enqueue_bundle_manifests(percent_complete, bundles_complete, total_bundles,
+                                                              f2f_futures, ff_futures, executor)
+
+        for future in concurrent.futures.as_completed(ff_futures):
+            f, bundle_uuid, bundle_version = future.result()
+            self.b2f[bundle_uuid + "." + bundle_version].remove(f["name"])
+            if len(self.b2f[bundle_uuid + "." + bundle_version]) == 0:
+                self.dispatch_callbacks(bundle_uuid, bundle_version, transformer, loader)
+        for bundle_uuid, bundle_version in self.staged_bundles:
+            self.dispatch_callbacks(bundle_uuid, bundle_version, transformer, loader)
+        logger.info("Done (%d/%d bundles)", bundles_complete, total_bundles)
+
+    def dispatch_callbacks(self, bundle_uuid, bundle_version, transformer, loader):
+        bundle_path = f"{self.sd}/bundles/{bundle_uuid}.{bundle_version}"
+        if transformer is not None:
+            tb = transformer(bundle_uuid=bundle_uuid, bundle_version=bundle_version, bundle_path=bundle_path)
+        if loader is not None:
+            loader(bundle=tb)
```

Comment:
Is `bundle` (`tb`) intended to be an object representation of the bundle? Asking out of curiousity - Matrix won't be using a per bundle loader. Instead, we'll need a final loader callback that probably only needs `staging_directory` as input.

Reply:
The semantics of the `tb` object are not defined yet. It's basically part of the contract between the transformer and the loader. We could put some constraints on it here, but I don't see a good reason to yet. Do you have any ideas?

Reply:
Me neither -  good as is